File: Dynamic-Programming/877. Stone Game/stoneGame.js
/**
 * @param {number[]} piles
 * @return {boolean}
 */

// Time complexity: O(n)
// Space complexity: O(n)
var stoneGame = function (piles) {
    let alex = 0;
    let lee = 0;

    while (piles.length) {
        alex += getMax(piles)[0];
        piles.splice(getMax(piles)[1], 1);

        lee += getMax(piles)[0];
        piles.splice(getMax(piles)[1], 1);
    }
    return alex > lee;
};

function getMax(piles) {
    let max = 0;
    let maxIndex = 0;
    for (let i = 0; i < piles.length; i++) {
        if (i === 0 || i === piles.length - 1) {
            max = Math.max(...piles);
            maxIndex = piles.indexOf(max);
        }
    }

    return [max, maxIndex]
}

File: Dynamic-Programming/877. Stone Game/stoneGame.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        def get_max(piles):
            max_index = 0
            max_num = 0
            for i in range(len(piles)):
                #tail or head
                if i == 0 or i == len(piles) - 1:
                    max_num = max(piles)
                    max_index = piles.index(max_num)
            return [max_num, max_index]
        
        alice = 0
        bob = 0
        while len(piles):    
            alice += get_max(piles)[0]
            piles.pop(get_max(piles)[1])
            bob += get_max(piles)[0]
            piles.pop(get_max(piles)[1])
            
        
        return alice > bob
            
                